src/LanguageServer/Protocol/Features/Options/SolutionAnalyzerConfigOptionsUpdater.cs

```diff
@@ -0,0 +1,273 @@
+    {
+        // Set of languages (keys of the dictionary) in the current snapshot that have their fallback options initialized from global options.
+        // The values are ignored.
+        private ImmutableDictionary<string, int> _initializedLanguages = ImmutableDictionary<string, int>.Empty;
+
+        // Ensures we only process one update at a time:
+        private readonly object _updateLock = new();
+
+        public void WorkspaceChanged(object? sender, WorkspaceChangeEventArgs args)
+        {
+            try
+            {
+                switch (args.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionChanged:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                    case WorkspaceChangeKind.ProjectAdded:
+                    case WorkspaceChangeKind.ProjectChanged:
+                    case WorkspaceChangeKind.ProjectReloaded:
+                    case WorkspaceChangeKind.ProjectRemoved:
+                        break;
+
+                    default:
+                        return;
+                }
+
+                if (args.OldSolution.SolutionState.ProjectCountByLanguage.KeysEqual(args.NewSolution.SolutionState.ProjectCountByLanguage))
+                {
+                    return;
+                }
+
+                UpdateSolution(
+                    transformation: solution => InitializeLanguages(solution, _initializedLanguages),
+                    onAfterUpdate: solution => _initializedLanguages = solution.SolutionState.ProjectCountByLanguage);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.Diagnostic))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
+
+        public void GlobalOptionsChanged(object? sender, OptionChangedEventArgs args)
+        {
+            try
+            {
+                if (!args.ChangedOptions.Any(static o => o.key.Option.Definition.IsEditorConfigOption))
+                {
+                    return;
+                }
+
+                UpdateSolution(
+                    transformation: solution => UpdateOptions(solution, args, _initializedLanguages),
+                    onAfterUpdate: null);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.Diagnostic))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
+
+        private void UpdateSolution(Func<Solution, Solution> transformation, Action<Solution>? onAfterUpdate)
+        {
+            var lockTaken = false;
+            try
+            {
+                // If another update is in progress wait until it completes.
+                Monitor.Enter(_updateLock, ref lockTaken);
```

Comment:
totally confused why using this pattern instead of literally just `lock (_updateLock)` :)

Reply:
oh.  i see you do something terrifying.